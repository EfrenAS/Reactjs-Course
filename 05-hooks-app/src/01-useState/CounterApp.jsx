import { useState } from 'react'

export default function CounterApp() {
  const [counters, setCounters] = useState({
    counter1: 10,
    counter2: 20,
    counter3: 30
  })

  const { counter1, counter2, counter3 } = counters

  function handleClick() {
    setCounters({...counters, counter1: counter1 + 1})
  }

  return ( 
    <>
      <h2>Counter1: {counter1} </h2>
      <h2>Counter2: {counter2} </h2>
      <h2>Counter3: {counter3} </h2>
      <br />
      <button onClick={handleClick} >+1</button>
    </>
  )
}