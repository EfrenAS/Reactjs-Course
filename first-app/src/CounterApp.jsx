import PropTypes from 'prop-types'
import { useState } from 'react'

export default function CounterApp ({ value }) {
  const [counter, setCounter] = useState(value)
  const handleAdd = () => setCounter(counter + 1)
  const hadleMinus = () => setCounter(counter - 1)
  const handleReset = () => setCounter(value)

  return (
    <>
      <h1>CounterApp</h1>
      <h2 data-testid='test-counter'>{counter}</h2>
      <div className='cr-content-buttons'>
        <button onClick={handleAdd}> +1 </button>
        <button onClick={hadleMinus}> -1 </button>
        <button aria-label='btn-reset' onClick={handleReset}> Reset </button>
      </div>
    </>
  )
}

CounterApp.propTypes = {
  value: PropTypes.number.isRequired
}
