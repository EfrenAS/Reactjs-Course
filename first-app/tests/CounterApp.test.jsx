import { render, screen, fireEvent } from "@testing-library/react"
import CounterApp from "../src/CounterApp.jsx"

describe("Pruebas en <CounterApp />", () => {
  const initialValue = 10
  const currentValue = 100

  test("Debe de hacer match con el snapshot", () => {
    const {container} = render(<CounterApp value={currentValue} />)

    expect(container).toMatchSnapshot()
  })

  test("Debe de mostrar el valor inicial de 100", () => {
    render(<CounterApp value={currentValue} />)

    expect(screen.getByText(currentValue)).toBeTruthy()
    expect(screen.getByTestId("test-counter").innerHTML).toContain(currentValue.toString())
  })

  test("Debe de incrementar el valor en 1", () => {
    render(<CounterApp value={initialValue} />)

    fireEvent.click(screen.getByText("+1"))
    expect(screen.getByText('11')).toBeTruthy()

  })

  test("Debe de decremnentar el valor en 1", () => {
    render(<CounterApp value={initialValue} />)

    fireEvent.click(screen.getByText("-1"))
    expect(screen.getByText('9')).toBeTruthy()

  })

  test("Debe de resetear el valor a 10", () => {
    render(<CounterApp value={initialValue} />)
    fireEvent.click(screen.getByText("+1"))
    fireEvent.click(screen.getByText("+1"))
    fireEvent.click(screen.getByText("+1"))

    //fireEvent.click(screen.getByText("Reset"))

    fireEvent.click(screen.getByRole("button", { name: "btn-reset" })) //aria-label='btn-reset' is a explcit label

    expect(screen.getByText('10')).toBeTruthy()
  })

})