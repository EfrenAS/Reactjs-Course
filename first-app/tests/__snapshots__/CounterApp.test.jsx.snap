// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <CounterApp /> Debe de hacer match con el snapshot 1`] = `
<div>
  <h1>
    CounterApp
  </h1>
  <h2
    data-testid="test-counter"
  >
    100
  </h2>
  <div
    class="cr-content-buttons"
  >
    <button>
       +1 
    </button>
    <button>
       -1 
    </button>
    <button
      aria-label="btn-reset"
    >
       Reset 
    </button>
  </div>
</div>
`;
